### ISSUE TITLE
[Bug] Material Visual has long paths making it difficult to publish to app store for iOS

### ISSUE THREAD

**Issue by mzhukovs** [xamarin/xamarin-macios#7208]
### Description

After archiving a Xamarin Forms iOS app, it is very troublesome to distribute it to the app store due to an issue with long paths which I believe stem from usage of the Xamarin.Forms.Visual.Material (apologies if I am mistaken):

Unable to copy file   "C:\Archives\2019-10-05\Cheshire.iOS 10-05-19 9.19   PM.xcarchive\Products\Applications\Cheshire.iOS.app\Frameworks\MaterialComponents.framework\MaterialActivityIndicator.bundle\Resources\en.lproj\MaterialActivityIndicator.strings"   to "obj\iPhone\Release\archives\Cheshire.iOS 10-05-19 9.19   PM.xcarchive\Products\Applications\Cheshire.iOS.app\Frameworks\MaterialComponents.framework\MaterialActivityIndicator.bundle\Resources\en.lproj\MaterialActivityIndicator.strings".   **The specified path, file name, or both are too long. The fully qualified file   name must be less than 260 characters, and the directory name must be less   than 248 characters.**

My project is located at C:\Repos\Cheshire\Cheshire.iOS , which is not too lengthy of a root. The only way I was able to get past this was by moving it and renaming to shorten all the way down to C:\c\i , which is a bit of a hassle. This is related to this VS issue but think it still makes sense to shorten the path for the framework if possible, because even if VS gets updated to allow for a custom path, the default would still be too long for most projects so shortening the resource path is still a good thing to do: https://developercommunity.visualstudio.com/content/problem/750967/cannot-create-ios-archive-process-cannot-be-execut.html 

--

### Steps to Reproduce

1. Create a Xamarin Forms project with iOS and Xamarin.Forms.Visual.Material package
2. Archive the iOS project
3. Attempt to Distribute it to App Store
4. Observe error as reported above regarding path being too long.

### Expected Behavior
Able to push build to the iOS App Store from VS2019 from a reasonable project directory.

### Actual Behavior
Unable to distribute to App Store because of paths being too long due to Material framework path lengths.

### Basic Information

- Version with issue:
- Last known good version: n/a
- IDE: VS2019 Enterprise latest
- Platform Target Frameworks: 
  - iOS
- Nuget Packages: Xamarin.Forms.Visual.Material

> VS bug [#1000573](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1000573)

**Comment by whitneyschmidt** [comment]
Thank you for your feedback!

Unfortunately this is a known issue on VS for Windows. Here are a couple workarounds, if you are still blocked after trying the following let us know.

1. As a workaround to unblock you, there is a command line tool you can use, which is called altool and you can find out how to use it in the Apple Documentation.

Something like:

`$ xcrun altool --upload-app -f ./path/to/my.ipa -t ios -u myappleid@example.com -p mypassword`

2. Shorten the path to your solution folder

3. It's not ideal, but if it's available to you, moving the solution to VS for Mac is an option

**Comment by mzhukovs** [comment]
In order to make this issue less likely to occur (at least for those using Xamarin Forms + Material Visual), is my original proposal of shortening the resource paths for the Xamarin.Forms.Visual.Material package something you would consider @StephaneDelcroix  ? That was the only thing preventing me from publishing, and forcing me to take option 2 above and moving my solution to a much shorter root.

**Comment by whitneyschmidt** [comment]
@mzhukovs Glad that you were able find a workaround to distribute your application!

Other users have encountered this issue, unfortunately we can't change the limitations of the Windows OS :(

I'm redirecting this issue to the external VS team because file path generation is not in the scope of our team's work.

**Comment by mzhukovs** [comment]
@whitneyschmidt thanks Whitney but I think there may be confusion around my original request here, which I have restated above - the issue that the publishing fails is because of just ONE of the packages that the app relies on, and that package is owned by Xamarin [Forms], where you should have control over things like filenames and output directories. My request was to shorten the resource paths for Xamarin.Forms.Visual.Material, specifically for the ActivityIndicator which has a very lengthy path (and which was causing my project to hit the error):

Frameworks\MaterialComponents.framework\MaterialActivityIndicator.bundle\Resources\en.lproj\MaterialActivityIndicator.strings

**Comment by Daddoon** [comment]
Must i open a new issue ?

This issue still exist on current Xamarin releases.
I cannot publish my app for my client because of this error.

I will try the given command line in this thread and test, but is it not possible to fix this somehow definitively ?

As XCode 11 removed Application Loader (thanks Apple...), publishing on Apple is very difficult.

**Comment by whitneyschmidt** [comment]
Update - @xamarin-release-manager is a bot! @Daddoon and @mzhukovs, this issue has been fixed and will be included in the next d16.4 preview 3.

The fix consists of:
Allowing to define a custom Archives Location path (as Android does) under Tools > Options > Xamarin iOS
Fixing the errors in the distribution workflow about MAX LENGTH violation

If you are still encountering this issue after updating, please feel free to reopen this issue.

**Comment by Daddoon** [comment]
Thanks for the feedback !

**Comment by llauderesv** [comment]
I'm encountering the same issue yesterday. But I did the workaround code below and it works!

`$ xcrun altool --upload-app -f ./path/to/my.ipa -t ios -u myappleid@example.com -p mypassword`

Thank you for this
